packages/vector_graphics/lib/src/listener.dart

```diff
@@ -621,21 +635,42 @@ class FlutterVectorGraphicsListener extends VectorGraphicsCodecListener {
     }
   }
 
+  int _createImageKey(int imageId, int format) {
+    return Object.hash(_id, imageId, format);
+  }
+
   @override
   void onImage(int imageId, int format, Uint8List data) {
     assert(format == 0); // Only PNG is supported.
-    _pendingImages.add(ImmutableBuffer.fromUint8List(data)
-        .then((ImmutableBuffer buffer) async {
-      final ImageDescriptor descriptor = await ImageDescriptor.encoded(buffer);
-      final Codec codec = await descriptor.instantiateCodec();
-      final FrameInfo info = await codec.getNextFrame();
-      final Image image = info.image;
-      buffer.dispose();
-      descriptor.dispose();
-      codec.dispose();
-      _images[imageId] = image;
-      return image;
-    }));
+    final Completer<void> completer = Completer<void>();
+    _pendingImages.add(completer.future);
+    final ImageStreamCompleter? cacheCompleter =
+        imageCache.putIfAbsent(_createImageKey(imageId, format), () {
+      return OneFrameImageStreamCompleter(ImmutableBuffer.fromUint8List(data)
+          .then((ImmutableBuffer buffer) async {
+        final ImageDescriptor descriptor =
+            await ImageDescriptor.encoded(buffer);
+        final Codec codec = await descriptor.instantiateCodec();
+        final FrameInfo info = await codec.getNextFrame();
+        final Image image = info.image;
+        buffer.dispose();
+        descriptor.dispose();
+        codec.dispose();
```

Comment:
make these nullable and put these in the `finally` of a try block?

Reply:
None of these should really throw, unless we give it bad data ... I guess in that case we could risk failing to dispose the buffer, so I'll move that into dispose, but otherwise if these throw there's probably not even anything to dispose of.